File: server/internal/crypto/wire_format_test.go
package crypto

import (
	"bytes"
	"testing"
)

// expectedSalts is the canonical byte content of every L9 salt. The Dart side
// test asserts the same literal byte sequences. If either side diverges,
// every cross language signature/HKDF derivation breaks silently
var expectedSalts = map[string][]byte{
	"vault-x3dh-v1":    SaltX3dh,
	"invite-v1":        SaltInvite,
	"seg-nonce-v1":     SaltSegNonce,
	"key-confirm-v1":   SaltKeyConfirm,
	"manifest-v1":      SaltManifest,
	"rotate-spk-v1":    SaltSpkRotate,
	"promote-v1":       SaltPromote,
	"join-complete-v1": SaltJoinComplete,
	"album-name-v1":    SaltAlbumNameHint,
}

func TestSalts_MatchExpectedAsciiBytes(t *testing.T) {
	for ascii, got := range expectedSalts {
		if !bytes.Equal(got, []byte(ascii)) {
			t.Errorf("salt %q: got %v, want ASCII bytes of literal", ascii, got)
		}
	}
}

func TestVerConstants_FixedByteValues(t *testing.T) {
	if VerAesGcm != 0x01 {
		t.Errorf("VerAesGcm = 0x%02x, want 0x01", VerAesGcm)
	}
	if VerChaPo != 0x02 {
		t.Errorf("VerChaPo = 0x%02x, want 0x02", VerChaPo)
	}
	if VerStreamGcm != 0x03 {
		t.Errorf("VerStreamGcm = 0x%02x, want 0x03", VerStreamGcm)
	}
}

func TestSegmentSize_OneMebibyte(t *testing.T) {
	if SegmentSize != 1048576 {
		t.Errorf("SegmentSize = %d, want 1048576 (1 MiB)", SegmentSize)
	}
}

func TestX3dhInfo_HappyPath(t *testing.T) {
	lkA := bytes.Repeat([]byte{0x11}, LkPubLen)
	lkB := bytes.Repeat([]byte{0x22}, LkPubLen)
	album := bytes.Repeat([]byte{0x33}, AlbumIDLen)

	got, err := X3dhInfo(lkA, lkB, album)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(got) != X3dhInfoLen {
		t.Fatalf("got len %d, want %d", len(got), X3dhInfoLen)
	}
	if !bytes.Equal(got[0:LkPubLen], lkA) {
		t.Error("first 32 bytes != lkA")
	}
	if !bytes.Equal(got[LkPubLen:2*LkPubLen], lkB) {
		t.Error("middle 32 bytes != lkB")
	}
	if !bytes.Equal(got[2*LkPubLen:X3dhInfoLen], album) {
		t.Error("last 16 bytes != album")
	}
}

func TestX3dhInfo_RejectsWrongSizes(t *testing.T) {
	good32 := make([]byte, 32)
	good16 := make([]byte, 16)

	for _, tc := range []struct {
		name        string
		a, b, album []byte
	}{
		{"lkA too short", make([]byte, 31), good32, good16},
		{"lkA too long", make([]byte, 33), good32, good16},
		{"lkB too short", good32, make([]byte, 31), good16},
		{"album too short", good32, good32, make([]byte, 15)},
		{"album too long", good32, good32, make([]byte, 17)},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := X3dhInfo(tc.a, tc.b, tc.album); err == nil {
				t.Error("expected error, got nil")
			}
		})
	}
}

File: frontend/test/crypto/wire_format_test.dart
import 'dart:typed_data';
import 'package:test/test.dart';
import 'package:keepsy/crypto/wire_format.dart';

void main() {
  group('L9 salts: expected ASCII bytes', () {
    final expected = <Uint8List, String>{
      kSaltX3dh: 'vault-x3dh-v1',
      kSaltInvite: 'invite-v1',
      kSaltSegNonce: 'seg-nonce-v1',
      kSaltKeyConfirm: 'key-confirm-v1',
      kSaltManifest: 'manifest-v1',
      kSaltSpkRotate: 'rotate-spk-v1',
      kSaltPromote: 'promote-v1',
      kSaltJoinComplete: 'join-complete-v1',
      kSaltAlbumNameHint: 'album-name-v1',
    };

    expected.forEach((salt, ascii) {
      test(ascii, () {
        final want = Uint8List.fromList(ascii.codeUnits);
        expect(salt, orderedEquals(want));
      });
    });
  });

  group('VER constants: fixed byte values', () {
    test('AES-GCM is 0x01', () => expect(kVerAesGcm, equals(0x01)));
    test('ChaCha-Poly is 0x02', () => expect(kVerChaPo, equals(0x02)));
    test('streaming GCM is 0x03', () => expect(kVerStreamGcm, equals(0x03)));
  });

  test('segment size is 1 MiB', () {
    expect(kSegmentSize, equals(1048576));
  });

  group('kX3dhInfoConstruction', () {
    test('layout: lkA || lkB || albumId, total 80 bytes', () {
      final lkA = Uint8List.fromList(List.filled(32, 0x11));
      final lkB = Uint8List.fromList(List.filled(32, 0x22));
      final album = Uint8List.fromList(List.filled(16, 0x33));

      final out = kX3dhInfoConstruction(lkA, lkB, album);

      expect(out.length, equals(80));
      expect(out.sublist(0, 32), orderedEquals(lkA));
      expect(out.sublist(32, 64), orderedEquals(lkB));
      expect(out.sublist(64, 80), orderedEquals(album));
    });

    test('rejects wrong-size lkA', () {
      expect(
        () => kX3dhInfoConstruction(
          Uint8List(31),
          Uint8List(32),
          Uint8List(16),
        ),
        throwsArgumentError,
      );
    });

    test('rejects wrong-size lkB', () {
      expect(
        () => kX3dhInfoConstruction(
          Uint8List(32),
          Uint8List(33),
          Uint8List(16),
        ),
        throwsArgumentError,
      );
    });

    test('rejects wrong-size albumId', () {
      expect(
        () => kX3dhInfoConstruction(
          Uint8List(32),
          Uint8List(32),
          Uint8List(15),
        ),
        throwsArgumentError,
      );
    });
  });
}
